### DISCUSSION TITLE
How to prevent data overfetching caused by circular populate

### DISCUSSION THREAD

**Discussion by alex996** [mikro-orm/mikro-orm#4196] [Q&A] [answered]
Let's say I have an `Author` that can have many `Book`s:

```ts
@Entity()
export class Author {
  @PrimaryKey()
  id!: number;

  @Property()
  name!: string;

  @OneToMany(() => Book, (book) => book.author)
  books = new Collection<Book>(this);
}

@Entity()
export class Book {
  @PrimaryKey()
  id!: number;

  @Property()
  title!: string;

  @ManyToOne(() => Author)
  author!: Author;
}
```

I set up an Express server like this:

```ts
app.get("/set", async (req, res) => {
  const a = new Author("Stephen King");
  const b1 = new Book("The Shining", a);
  const b2 = new Book("Misery", a);
  const b3 = new Book("The Green Mile", a);

  await orm.em.persistAndFlush([b1, b2, b3]);

  res.json({});
});

app.get("/get", async (req, res) => {
  const b = await orm.em.findOneOrFail(Book, 1, {
    populate: ["author", "author.books"],
  });

  res.json(b);
});
```

I first run `curl localhost:3000/set` and then `curl localhost:3000/get`. I expect to get the JSON data in this shape:

```
book
  author
    books
```

However, I actually get:

```
book
  author
    books
      author
```

```jsonc
{
  "id": 1,
  "title": "The Shining",
  "author": {
    "id": 1,
    "name": "Stephen King",
    "books": [
      {
        "id": 1,
        "title": "The Shining"
      },
      {
        "id": 2,
        "title": "Misery",
        "author": { // <----- This is being sent unexpectedly
          "id": 1,
          "name": "Stephen King"
        }
      },
      {
        "id": 3,
        "title": "The Green Mile",
        "author": { // <----- this too
          "id": 1,
          "name": "Stephen King"
        }
      }
    ]
  }
}
```

The payload includes extra data: `book.author.books.author`. It's puzzling because

```ts
// our populate hints are:
populate: ["author", "author.books"],
// and not
populate: ["author", "author.books", "author.books.author"],
```

### Problem

This leads to duplicate data and much larger payload size, i.e. overfetching. I have a project with many more circular relationships like this. If I am not careful, my response body can easily balloon to 1-3 MB.

### Question

Does MikroORM offer any tools to mitigate this? A global config setting? A different way of serializing?

I realize that I could write a recursive function that walks a loaded entity and serializes it strictly according to the populate hints array. Is this the way to go or I am doing something fundamentally wrong?

Thank you.

**Comment by B4nan** [comment, marked as answer]
https://mikro-orm.io/docs/serializing#explicit-serialization

**Reply by B4nan** [reply]
And subscribe here for more updates: #4138

**Reply by alex996** [reply]
Fantastic, thank you!
